File: vortexdbg-api/src/main/java/com/vortexdbg/debugger/FunctionCallListener.java
package com.vortexdbg.debugger;

import com.vortexdbg.Emulator;
import com.vortexdbg.arm.FunctionCall;

public abstract class FunctionCallListener {

    public abstract void onCall(Emulator<?> emulator, long callerAddress, long functionAddress);

    public abstract void postCall(Emulator<?> emulator, long callerAddress, long functionAddress, Number[] args);

    public void onDebugPushFunction(Emulator<?> emulator, FunctionCall call) {
    }
    public void onDebugPopFunction(Emulator<?> emulator, long address, FunctionCall call) {
    }

}

File: vortexdbg-api/src/main/kotlin/com/vortexdbg/debugger/FunctionCallListener.kt
package com.vortexdbg.debugger

import com.vortexdbg.Emulator
import com.vortexdbg.arm.FunctionCall

abstract class FunctionCallListener {

    abstract fun onCall(emulator: Emulator<*>, callerAddress: Long, functionAddress: Long)

    abstract fun postCall(emulator: Emulator<*>, callerAddress: Long, functionAddress: Long, args: Array<Number>)

    open fun onDebugPushFunction(emulator: Emulator<*>, call: FunctionCall) {
    }

    open fun onDebugPopFunction(emulator: Emulator<*>, address: Long, call: FunctionCall) {
    }

}
